src/cargo/core/resolver/context.rs

```diff
@@ -135,8 +107,13 @@ impl ResolverContext {
                 // versions came from a `[patch]` source.
                 if let Some((_, dep)) = parent {
                     if dep.source_id() != id.source_id() {
-                        let key = (id.name(), dep.source_id(), id.version().into());
-                        let prev = self.activations.insert(key, (summary.clone(), age));
+                        let new_id =
+                            PackageId::new(id.name(), id.version().clone(), dep.source_id());
```

Comment:
Note for self: This call to `PackageId::new` is a hazard for parallelism. It would still be a hazard if we directly called `ActivationKey::into`. This is in the hottest part of the resolver constructing a new lock/cached object. But, this case only occurs when we are dealing with a patched source. This is not a problem for my benchmarking tool because I don't deal with patched sources. This is not a problem for cargo because it does not run in parallel.